projects/plugins/protect/src/class-status.php

```diff
@@ -294,4 +296,100 @@ public static function delete_option() {
 		delete_option( self::OPTION_TIMESTAMP_NAME );
 	}
 
+	/**
+	 * Prepare the report data for the UI
+	 *
+	 * @param string $body The report status report response.
+	 * @return object The normalized report data.
+	 */
+	private static function normalize_report_data( $body ) {
+		$report_data = json_decode( $body );
+
+		$installed_plugins    = Plugins_Installer::get_plugins();
+		$report_data->plugins = self::merge_installed_and_checked_lists( $installed_plugins, $report_data->plugins, array( 'type' => 'plugin' ) );
+
+		$installed_themes    = Sync_Functions::get_themes();
+		$report_data->themes = self::merge_installed_and_checked_lists( $installed_themes, $report_data->themes, array( 'type' => 'theme' ) );
+
+		$report_data->core = self::normalize_core_information( $report_data->core );
+
+		return $report_data;
+	}
+
+	/**
+	 * Merges the list of installed extensions with the list of extensions that were checked for known vulnerabilities and return a normalized list to be used in the UI
+	 *
+	 * @param object $installed The list of installed extensions, where each attribute key is the extension slug.
+	 * @param object $checked   The list of checked extensions.
+	 * @param array  $append    Additional data to append to each result in the list.
+	 * @return array Normalized list of extensions.
+	 */
+	private static function merge_installed_and_checked_lists( $installed, $checked, $append ) {
+		$new_list = array();
+		foreach ( $installed as $slug => $item ) {
+			if ( isset( $checked->{ $slug } ) && $checked->{ $slug }->version === $installed[ $slug ]['Version'] ) {
+				array_push(
+					$new_list,
+					(object) array_merge(
+						array(
+							'name'            => $installed[ $slug ]['Name'],
+							'version'         => $checked->{ $slug }->version,
+							'vulnerabilities' => $checked->{ $slug }->vulnerabilities,
+							'not_checked'     => false,
+						),
+						$append
+					)
+				);
+			} else {
+				array_push(
+					$new_list,
+					(object) array_merge(
+						array(
+							'name'            => $installed[ $slug ]['Name'],
+							'version'         => $installed[ $slug ]['Version'],
+							'vulnerabilities' => array(),
+							'not_checked'     => true,
+						),
+						$append
+					)
+				);
+			}
+		}
+		usort(
+			$new_list,
+			function ( $a, $b ) {
+				$vuls_a    = count( $a->vulnerabilities ) > 0 ? 2 : 0;
+				$vuls_b    = count( $b->vulnerabilities ) > 0 ? 2 : 0;
+				$checked_a = $a->not_checked ? 1 : 0;
+				$checked_b = $b->not_checked ? 1 : 0;
+				return $vuls_b + $checked_b - ( $vuls_a + $checked_a );
```

Comment:
I have to admit I'm not sure I fully got behind what is happening here, but if I am not mistaken, we want to have the ones that have vulnerabilities at the top and then the ones that have been checked, and the rest below; did I get that right?

And if so, what would you think about this alternative (imho simpler) approach?

```suggestion
				$vulns_count_diff = count( $a->vulnerabilities ) <=> count( $b->vulnerabilities );
				if ( 0 !== $vulns_count_diff ) {
				    return $vulns_count_diff;
				}
				return $b->not_checked <=> $a->not_checked;
```

Note: I have not checked if $a or $b should go on the left-hand or right-hand side of the spaceship operator here, but I think it should be this way around as long as we keep the `not_checked` and are not changing it to `checked` (see my previous comment).

Edit: Forgot we have to support below PHP7 here for now; however, the same can be achieved (and imho would still be simpler) without it.

Reply:
Interesting! Disappointing we can't use the spaceship operator for now, that is a super clean implementation. I do believe this approach might change the logic of the sort a bit, depending on how the list is arranged by default. 

With the current approach, it only considers the presence of vulnerabilities (vulnerabilities.length > 0). Each section of the result (has vulnerabilities, not checked, has no vulnerabilities) is left alphabetical. 

This suggested approach directly compares the amount of vulnerabilities of each item against each other, pushing the items with the most vulnerabilities to the very top of each section. Is this part of the suggestion?

Reply:
> Is this part of the suggestion?

Aha! It was actually not, I just had a hard time grasping what that function exactly did; which was also the reason why I tried to suggest a different approach :) 

So if I see it correctly, if we make the one line (assuming we could use the spaceship operator, which we cannot) like this:

```php
$vulns_count_diff = count( $a->vulnerabilities ) > 0 ? 1 : 0 <=> count( $b->vulnerabilities ) > 0 ? 1 : 0;
```

would make it the same again, huh? Not sure if I still find this easier to read.

An alternative would be

```php
$vulns_count_diff = ! empty( $a->vulnerabilities ) <=> ! empty( $b->vulnerabilities );
```

But then again we can not really use the spaceship operator, sadly. 

Maybe it is just my brain being weird, but what I found most complicated to understand is the `return ...` line. Basically, if we can make this two or more simple steps rather than checking all the things and "calculating the result together", I'd find it way more accessible regarding understanding. 🙂

Reply:
How about super simple? More lines of code, easier to read and understand.

```php
usort(
	$new_list,
	function ( $a, $b ) {
		// sort primarily based on the presence of vulnerabilities
		if ( ! empty( $a->vulnerabilities ) && empty( $b->vulnerabilities ) ) {
			return -1;
		}
		if ( empty( $a->vulnerabilities ) && ! empty( $b->vulnerabilities ) ) {
			return 1;
		}
		// sort secondarily on whether the item has been checked
		if ( $a->checked && ! $b->checked ) {
			return 1;
		}
		if ( ! $a->checked && $b->checked ) {
			return -1;
		}

		return 0;
	}
);
```